### PR TITLE
OCPQE-13027: Retry when Polarion failed to create test run

### PR DESCRIPTION
**Pull request by liangxia**
When Polarion succeed to create test run, work as previous, no retry.
```
bash-4.4# tools/polarshift.rb create-run -f /67_IPI\ on\ Nutanix\ \&\ FIPS\ on\ \&\ OVN\ \&\ CSI\ pvc.test_run.yaml --no-wait 
[09:17:21] INFO> HTTP POST user@https://POLARSHIFT-URL/project/OSE/run
[09:17:22] INFO> HTTP POST took 1.055 sec: 202 Accepted | application/json 267 bytes
[09:17:22] INFO> to check operation status manually, you can: curl 'https://POLARSHIFT-URL/polarion/request/63847ca2a0de9600108de648' -u user:thepassword
waiting for operation up to 360 seconds..
[09:17:22] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847ca2a0de9600108de648
[09:17:23] INFO> HTTP GET took 1.049 sec: 200 OK | application/json 92 bytes
[09:17:38] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847ca2a0de9600108de648
[09:17:39] INFO> HTTP GET took 1.050 sec: 200 OK | application/json 581 bytes
test run id: '20221128-0917'
```

When Polarion failed to create test run, retry until succeed to create the test run, or fail if we have retried 5 times.
```
bash-4.4# tools/polarshift.rb create-run -f /67_IPI\ on\ Nutanix\ \&\ FIPS\ on\ \&\ OVN\ \&\ CSI\ pvc.test_run.yaml --no-wait 
[09:18:23] INFO> HTTP POST user@https://POLARSHIFT-URL/project/OSE/run
[09:18:24] INFO> HTTP POST took 1.054 sec: 202 Accepted | application/json 267 bytes
[09:18:24] INFO> to check operation status manually, you can: curl 'https://POLARSHIFT-URL/polarion/request/63847ce0ebeb14000fb837de' -u user:thepassword
waiting for operation up to 360 seconds..
[09:18:24] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847ce0ebeb14000fb837de
[09:18:25] INFO> HTTP GET took 1.049 sec: 200 OK | application/json 122 bytes
[09:18:40] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847ce0ebeb14000fb837de
[09:18:41] INFO> HTTP GET took 1.064 sec: 200 OK | application/json 5766 bytes
[09:18:41] WARN> PolarShift operation failed:
#<RuntimeError: no corresponding PolarShift status for ''>
    /opt/app-root/src/app/models/test_record.rb:69:in `polarion_to_polarshift_status'
    /opt/app-root/src/app/models/test_record.rb:12:in `initialize'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:369:in `new'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:369:in `block in create_test_run'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:368:in `map'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:368:in `create_test_run'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:48:in `process_queue_item'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:37:in `process_op_queue'
    /opt/app-root/src/lib/tasks/polarion.rake:5:in `block (2 levels) in <top (required)>'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:250:in `block in execute'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:250:in `each'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:250:in `execute'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:194:in `block in invoke_with_call_chain'
    /opt/rh/rh-ruby26/root/usr/share/ruby/monitor.rb:230:in `mon_synchronize'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:187:in `invoke_with_call_chain'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:180:in `invoke'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:152:in `invoke_task'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:108:in `block (2 levels) in top_level'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:108:in `each'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:108:in `block in top_level'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:117:in `run_with_threads'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:102:in `top_level'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands/rake_proxy.rb:14:in `block in run_rake_task'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:178:in `standard_exception_handling'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands/rake_proxy.rb:11:in `run_rake_task'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands/commands_tasks.rb:51:in `run_command!'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands.rb:18:in `<top (required)>'
    /opt/app-root/src/bin/rails:9:in `require'
    /opt/app-root/src/bin/rails:9:in `<main>'
[09:18:41] INFO> HTTP POST user@https://POLARSHIFT-URL/project/OSE/run
[09:18:42] INFO> HTTP POST took 1.053 sec: 202 Accepted | application/json 3106 bytes
[09:18:42] INFO> to check operation status manually, you can: curl 'https://POLARSHIFT-URL/polarion/request/63847cf2a0de9600108de649' -u user:thepassword
waiting for operation up to 360 seconds..
[09:18:42] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847cf2a0de9600108de649
[09:18:43] INFO> HTTP GET took 1.051 sec: 200 OK | application/json 2961 bytes
[09:18:58] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847cf2a0de9600108de649
[09:18:59] INFO> HTTP GET took 1.042 sec: 200 OK | application/json 8606 bytes
[09:18:59] WARN> PolarShift operation failed:
#<RuntimeError: no corresponding PolarShift status for ''>
    /opt/app-root/src/app/models/test_record.rb:69:in `polarion_to_polarshift_status'
    /opt/app-root/src/app/models/test_record.rb:12:in `initialize'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:369:in `new'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:369:in `block in create_test_run'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:368:in `map'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:368:in `create_test_run'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:48:in `process_queue_item'
    /opt/app-root/src/app/controllers/concerns/polarion.rb:37:in `process_op_queue'
    /opt/app-root/src/lib/tasks/polarion.rake:5:in `block (2 levels) in <top (required)>'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:250:in `block in execute'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:250:in `each'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:250:in `execute'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:194:in `block in invoke_with_call_chain'
    /opt/rh/rh-ruby26/root/usr/share/ruby/monitor.rb:230:in `mon_synchronize'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:187:in `invoke_with_call_chain'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/task.rb:180:in `invoke'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:152:in `invoke_task'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:108:in `block (2 levels) in top_level'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:108:in `each'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:108:in `block in top_level'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:117:in `run_with_threads'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:102:in `top_level'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands/rake_proxy.rb:14:in `block in run_rake_task'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/rake-12.0.0/lib/rake/application.rb:178:in `standard_exception_handling'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands/rake_proxy.rb:11:in `run_rake_task'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands/commands_tasks.rb:51:in `run_command!'
    /opt/app-root/src/bundle/ruby/2.6.0/gems/railties-5.0.5/lib/rails/commands.rb:18:in `<top (required)>'
    /opt/app-root/src/bin/rails:9:in `require'
    /opt/app-root/src/bin/rails:9:in `<main>'
[09:18:59] INFO> HTTP POST user@https://POLARSHIFT-URL/project/OSE/run
[09:19:00] INFO> HTTP POST took 1.050 sec: 202 Accepted | application/json 5946 bytes
[09:19:00] INFO> to check operation status manually, you can: curl 'https://POLARSHIFT-URL/polarion/request/63847d046bd97a000f53f32d' -u user:thepassword
waiting for operation up to 360 seconds..
[09:19:00] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847d046bd97a000f53f32d
[09:19:01] INFO> HTTP GET took 1.049 sec: 200 OK | application/json 5771 bytes
[09:19:16] INFO> HTTP GET user@https://POLARSHIFT-URL/polarion/request/63847d046bd97a000f53f32d
[09:19:18] INFO> HTTP GET took 1.047 sec: 200 OK | application/json 6260 bytes
test run id: '20221128-0919'
```

### CODE REVIEW

File: lib/polarshift/request.rb [hunk 1]
```diff
@@ -176,14 +176,19 @@ def create_run(project_id:,
...
-        if res[:exitstatus] == 202
-          op_url = JSON.load(res[:response])["operation_result_url"]
-          logger.info "to check operation status manually, you can: " \
-            "curl '#{op_url}' -u user:thepassword"
-          pr = wait_op(url: op_url, timeout: timeout)
-        else
-          raise %Q{got status "#{res[:exitstatus]}" creating a new test run:\n#{res[:response]}}
+        for retries in 1..5 do
+          res = create_run(**opts)
+          if res[:exitstatus] == 202
```
Comment (jhou1) [inline lib/polarshift/request.rb:181]: Thank you @liangxia , can you elaborate what `202` represents here and why we raise an exception if it's not `202`?

Comment (liangxia) [reply lib/polarshift/request.rb:181]:
```
[09:27:00] INFO> HTTP POST user@https://POLARSHIFT-URL/project/OSE/run
[09:27:02] INFO> HTTP POST took 1.086 sec: 202 Accepted | application/json 267 bytes
```
This is the `202` comes from.

Comment (jhou1) [reply lib/polarshift/request.rb:181]: OK, that seems to be the original code here.

File: lib/polarshift/request.rb [hunk 2]
```diff
@@ -374,7 +379,8 @@ def wait_op(url: nil, id: nil, timeout: 360)
           when "queued", "running"
             next
           when "failed"
-            raise "PolarShift operation failed:\n#{res["error"]}"
+            logger.warn "PolarShift operation failed:\n#{res["error"]}"
```
Comment (jhou1) [inline lib/polarshift/request.rb:382]: what about using an error indicator than warning?

File: lib/polarshift/request.rb (hunk 1 above)
Comment (liangxia) [reply lib/polarshift/request.rb:181]: It's explained in https://gitlab.cee.redhat.com/aosqe/polarshift/-/blob/master/doc/API.adoc#create-test-run

File: lib/polarshift/request.rb (hunk 2 above)
Comment (liangxia) [reply lib/polarshift/request.rb:382]: It's indeed an error status ( a `failed` status from Polarion) initially.
As we are retrying when met this, and it's very likely to pass within 5 retries based my testing locally. So we finally passed this steps and the CI move on. I think a warn would be more appropriate here.
Of course, an error here is also acceptable to me. Let me know which one do you prefer ?

Comment (jhou1) [reply lib/polarshift/request.rb:382]: OK, thanks for the clarification

**Comment by pruan-rht** [comment]
we might want to put a little sleep between the retries.  I'm not sure doing a 5 tries consecutively is long enough

**Pull request merged**